From a2c8327a1b534574d51f5ea42009876e0008efd2 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.com>
Date: Mon, 9 Sep 2024 17:28:00 +0100
Subject: [PATCH 1346/1350] drm/vc4: Do not include writeback conn load in load
 tracker

The transposer/writeback connector should be running with a
lower priority, so shouldn't be factored into the load
calculations.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>
---
 drivers/gpu/drm/vc4/vc4_kms.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

--- a/drivers/gpu/drm/vc4/vc4_kms.c
+++ b/drivers/gpu/drm/vc4/vc4_kms.c
@@ -674,17 +674,26 @@ static int vc4_load_tracker_atomic_check
 	for_each_oldnew_plane_in_state(state, plane, old_plane_state,
 				       new_plane_state, i) {
 		struct vc4_plane_state *vc4_plane_state;
+		struct vc4_crtc *vc4_crtc;
 
 		if (old_plane_state->fb && old_plane_state->crtc) {
 			vc4_plane_state = to_vc4_plane_state(old_plane_state);
-			load_state->membus_load -= vc4_plane_state->membus_load;
-			load_state->hvs_load -= vc4_plane_state->hvs_load;
+			vc4_crtc = to_vc4_crtc(old_plane_state->crtc);
+
+			if (!vc4_crtc->feeds_txp) {
+				load_state->membus_load -= vc4_plane_state->membus_load;
+				load_state->hvs_load -= vc4_plane_state->hvs_load;
+			}
 		}
 
 		if (new_plane_state->fb && new_plane_state->crtc) {
 			vc4_plane_state = to_vc4_plane_state(new_plane_state);
-			load_state->membus_load += vc4_plane_state->membus_load;
-			load_state->hvs_load += vc4_plane_state->hvs_load;
+			vc4_crtc = to_vc4_crtc(new_plane_state->crtc);
+
+			if (!vc4_crtc->feeds_txp) {
+				load_state->membus_load += vc4_plane_state->membus_load;
+				load_state->hvs_load += vc4_plane_state->hvs_load;
+			}
 		}
 	}
 
